### PR TITLE
xtensa RzIL

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

xtensa: add rzil support

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by b1llow** [comment]
Basic pass is done.

File: librz/include/rz_il/rz_il_opbuilder_begin.h [hunk 1]
```diff
@@ -82,6 +82,7 @@
 #define FROUND(rmode, fl)          rz_il_op_new_fround(rmode, fl)
 #define FSQRT(rmode, fl)           rz_il_op_new_fsqrt(rmode, fl)
 #define FRSQRT(rmode, fl)          rz_il_op_new_frsqrt(rmode, fl)
+#define FEXCEPT(e, fl)             rz_il_op_new_fexcept(e, fl)
```
Comment (XVilka) [inline librz/include/rz_il/rz_il_opbuilder_begin.h]: Seems there is mismatch - in some places it's `except`, in others - `expect`

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by b1llow** [comment]
Still debugging with emulateme, shouldn't take long!

**Review by XVilka — COMMENTED** [review]
Could you please extract RzIL VM changes into a separate PR? After you finish the "emulateme" test, of course.

**Comment by b1llow** [comment]
the emulate-me test is done.

- https://github.com/rizinorg/rizin/pull/4771
- https://github.com/rizinorg/rizin/pull/4772

**Comment by XVilka** [comment]
@imbillow please rebase (without changes to the `rz-test -y` since they will be merged separately).

File: librz/include/rz_il/rz_il_events.h [hunk 2]
```diff
@@ -68,9 +70,9 @@ typedef struct rz_il_vm_event_t {
 
 RZ_API RZ_OWN RzILEvent *rz_il_event_exception_new(RZ_NONNULL const char *exception);
 RZ_API RZ_OWN RzILEvent *rz_il_event_pc_write_new(RZ_NONNULL const RzBitVector *old_pc, RZ_NONNULL const RzBitVector *new_pc);
-RZ_API RZ_OWN RzILEvent *rz_il_event_mem_read_new(RZ_NONNULL const RzBitVector *addr, RZ_NULLABLE const RzBitVector *value);
+RZ_API RZ_OWN RzILEvent *rz_il_event_mem_read_new(RzILMemIndex index, RZ_NONNULL const RzBitVector *addr, RZ_NULLABLE const RzBitVector *value);
```
Comment (XVilka) [inline librz/include/rz_il/rz_il_events.h]: Please extract these changes into a separate PR again.

**Review by XVilka — CHANGES_REQUESTED** [review]
UBSAN problem:
```
[XX] 833 ms   db/asm/xtensa <asm> retw.n
-- <asm> retw.n <--- 1df0 ---> <IL>
-- disasm exit status: 0
-- IL exit status: 0
-- IL stderr
../librz/arch/isa/xtensa/xtensa_il.c:1323:9: runtime error: left shift of 3 by 30 places cannot be represented in type 'int'


[XX] 803 ms   db/asm/xtensa <asm> retw
-- <asm> retw <--- 900000 ---> <IL>
-- disasm exit status: 0
-- IL exit status: 0
-- IL stderr
../librz/arch/isa/xtensa/xtensa_il.c:1323:9: runtime error: left shift of 3 by 30 places cannot be represented in type 'int'
```

File: librz/il/il_events.c [hunk 3]
```diff
@@ -94,10 +94,11 @@ RZ_API RZ_OWN RzILEvent *rz_il_event_pc_write_new(RZ_NONNULL const RzBitVector *
 
 /**
  * Creates an RzILEvent of type RZ_IL_EVENT_MEM_READ
+ * \param index, RzILMemIndex, index of the memory to read
  * \param addr, RzBitVector, address of the memory where the read op has occurred
  * \param value, RzBitVector, value read from the variable
  */
-RZ_API RZ_OWN RzILEvent *rz_il_event_mem_read_new(RZ_NONNULL const RzBitVector *address, RZ_NULLABLE const RzBitVector *value) {
+RZ_API RZ_OWN RzILEvent *rz_il_event_mem_read_new(RzILMemIndex index, RZ_NONNULL const RzBitVector *address, RZ_NULLABLE const RzBitVector *value) {
```
Comment (XVilka) [inline librz/il/il_events.c]: @Rot127 @thestr4ng3r take a look at this change please

**Comment by b1llow** [comment]
#4789

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 4]
```diff
@@ -0,0 +1,2178 @@
...
+		{ PS_RING, 6, 2 },
+		{ PS_OWB, 8, 4 },
+		{ PS_CALLINC, 16, 2 },
+		{ PS_WOE, 18, 1 },
+	},
+	.size = 7,
+	.width = 32,
+};
+
+// static const RegFieldTbl fcr_field_tbl = {
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c]: Is this unused code? Why keep it, do you plan to use/uncomment it in the future?

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 5]
```diff
@@ -0,0 +1,2178 @@
...
+		eff = f_cons_(eff, FSR_field_set(FSR_U, B2U32(FEXCEPT(RZ_FLOAT_E_UNDERFLOW, VARL("fsr_v")))));
+	}
+	if (fs & FSR_Z) {
+		eff = f_cons_(eff, FSR_field_set(FSR_Z, B2U32(FEXCEPT(RZ_FLOAT_E_DIV_ZERO, VARL("fsr_v")))));
+	}
+	rz_warn_if_fail(eff);
+	return eff;
+}
+
+static RzILOpEffect *ARi_set(RzILOpPure *i, RzILOpPure *v) {
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c:241]: Is it possible to make it less nested somehow to not break formatting like that?

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 6]
```diff
@@ -0,0 +1,2178 @@
...
+
+static RzAnalysisLiftedILOp op_binary4(XtensaContext *ctx, fn_op2 f) {
+	rz_return_val_if_fail(FORMAT == XTENSA_INSN_FORM_RRR && RRR_s(ctx) % 4 == 0, NULL);
+	return SETG(REGN(0), f(f(f(IREGi(1, 0), IREGi(1, 1)), IREGi(1, 2)), IREGi(1, 3)));
+}
+
+static RzAnalysisLiftedILOp op_binary8(XtensaContext *ctx, fn_op2 f) {
+	rz_return_val_if_fail(FORMAT == XTENSA_INSN_FORM_RRR && RRR_s(ctx) % 8 == 0, NULL);
+	return SETG(REGN(0),
+		f(f(f(f(f(f(f(IREGi(1, 0), IREGi(1, 1)), IREGi(1, 2)), IREGi(1, 3)), IREGi(1, 4)), IREGi(1, 5)), IREGi(1, 6)), IREGi(1, 7)));
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c]: Split this line, please

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 7]
```diff
@@ -0,0 +1,2178 @@
...
+static RzAnalysisLiftedILOp op_nop(XtensaContext *ctx) {
+	return NOP();
+}
+
+static RzAnalysisLiftedILOp op_nsa(XtensaContext *ctx) {
+	return SEQ3(
+		SETL("ars", IREG(1)),
+		SETL("sign", EXTRACT32(VARL("ars"), U32(31), U32(1))),
+		SETG(REGN(0),
+			ITE(EQ(VARL("sign"), EXTRACT32(VARL("ars"), U32(0), U32(31))), U32(31),
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c]: Same issue

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 8]
```diff
@@ -0,0 +1,2178 @@
...
+																    BOOL_TO_BV(VARLP("b0"), 32))))),
+													U32(1)))))))))))));
+}
+
+static RzAnalysisLiftedILOp op_nsau(XtensaContext *ctx) {
+	return SEQ3(
+		SETL("ars", IREG(1)),
+		SETL("sign", U32(0)),
+		SETG(REGN(0),
+			ITE(EQ(VARL("sign"), VARL("ars")), U32(32),
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c]: And here

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 9]
```diff
@@ -0,0 +1,2178 @@
...
+		SETG(REGN(0), SHIFTL0(IREG(1), VARL("sa"))));
+}
+
+static RzAnalysisLiftedILOp op_slli(XtensaContext *ctx) {
+	return SEQ2(
+		SETL("sa", U32(IMM(2))),
+		SETG(REGN(0), SHIFTL0(IREG(1), VARL("sa"))));
+}
+
+// TODO: see Divide and Square Root Sequences on page 110.
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c:1528]: Do you plan to fix this?

File: librz/arch/isa/xtensa/xtensa_il.c [hunk 10]
```diff
@@ -0,0 +1,2178 @@
...
+		SETG(REGN(0), F2INT(32, RZ_FLOAT_RMODE_RNA, VARL("fres"))),
+		FSR_set(FSR_V | FSR_I, VARL("fres")));
+}
+
+// TODO: interrupt
+static RzAnalysisLiftedILOp op_waiti(XtensaContext *ctx) {
+	return NOP();
+}
+
+// TODO: datatlb
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa_il.c:1692]: I think these todos aren't needed. There is no point of uplifting barrier instructions to RzIL, I think.

**Review by XVilka — CHANGES_REQUESTED** [review]
Thanks, impressive work as always. Please rebase and address these few nitpicks.

**Comment by XVilka** [comment]
```
../librz/arch/isa/xtensa/xtensa_il.c:1220:7: error: expected expression
                                                LET("b2", EQ(VARL("sign"), EXTRACT32(VARLP("t2"), U32(4), U32(4))), ))))));
                                                ^
../librz/include/rz_il/rz_il_opbuilder_begin.h:160:58: note: expanded from macro 'LET'
#define LET(name, v, body) rz_il_op_new_let(name, v, body)
                                                         ^
1 error generated.
```

File: librz/arch/isa/xtensa/xtensa_il.c (hunk 9 above)
Comment (b1llow) [reply librz/arch/isa/xtensa/xtensa_il.c:1528]: I'm planning to do this in another new PR, and the instructions are so strange that I'm not sure it's necessary to implement them.

image

Comment (XVilka) [reply librz/arch/isa/xtensa/xtensa_il.c:1528]: It depends if these instructions are used in the real code or not. Yep, it's okay to do in a subsequent PR.

**Review by XVilka — APPROVED** [review]
